ethexe/runtime/common/src/state.rs

```diff
@@ -812,12 +827,81 @@ impl Mailbox {
     }
 }
 
-#[derive(Clone, Default, Debug, Encode, Decode, PartialEq, Eq, derive_more::Into)]
-#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, derive_more::Into)]
 pub struct MemoryPages(MemoryPagesInner);
 
-// TODO (breathx): replace BTreeMap here with fixed size array.
-pub type MemoryPagesInner = BTreeMap<RegionIdx, HashOf<MemoryPagesRegion>>;
+impl Default for MemoryPages {
+    fn default() -> Self {
+        Self([MaybeHashOf::empty(); 256])
+    }
+}
+// implemented manually because serde_derive does not work with arrays larger than 32 elements
+#[cfg(feature = "std")]
+impl serde::Serialize for MemoryPages {
+    fn serialize<S>(&self, serializer: S) -> core::result::Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut seq = serializer.serialize_tuple(self.0.len())?;
+        for elem in &self.0[..] {
+            seq.serialize_element(elem)?;
+        }
+
+        seq.end()
+    }
+}
+
+#[cfg(feature = "std")]
+impl<'a> serde::Deserialize<'a> for MemoryPages {
+    fn deserialize<D>(deserializer: D) -> core::result::Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'a>,
+    {
+        struct ArrayVisitor {}
+
+        impl<'de> Visitor<'de> for ArrayVisitor {
+            type Value = MemoryPagesInner;
+
+            fn expecting(&self, formatter: &mut alloc::fmt::Formatter) -> alloc::fmt::Result {
+                formatter.write_str("an array of length 256")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> core::result::Result<Self::Value, A::Error>
+            where
+                A: serde::de::SeqAccess<'de>,
+            {
+                let mut arr = [MaybeHashOf::empty(); 256];
+                for (i, hash) in arr.iter_mut().enumerate() {
+                    *hash = seq
+                        .next_element()?
+                        .ok_or_else(|| serde::de::Error::invalid_length(i, &self))?;
+                }
+                Ok(arr)
+            }
+        }
+
+        let visitor = ArrayVisitor {};
+        deserializer.deserialize_tuple(256, visitor).map(Self)
+    }
+}
+
+impl Index<RegionIdx> for MemoryPages {
+    type Output = MaybeHashOf<MemoryPagesRegion>;
+
+    fn index(&self, idx: RegionIdx) -> &Self::Output {
+        &self.0[idx.0 as usize]
+    }
+}
+
+impl IndexMut<RegionIdx> for MemoryPages {
+    fn index_mut(&mut self, index: RegionIdx) -> &mut Self::Output {
+        &mut self.0[index.0 as usize]
+    }
+}
+
+/// An inner structure for [`MemoryPages`]. Has at most 256 entries because
+/// [`RegionIdx`] is `u8` internally.
+pub type MemoryPagesInner = [MaybeHashOf<MemoryPagesRegion>; 256];
```

Comment:
Why we should support all 256 ? We use only 16 regions , so what the reason to make it 256 size which increase encode/decode serialize/deserialize time significally